src/engine.rs

```diff
@@ -141,344 +161,250 @@ impl FileEngineInner {
-            }
+    fn first_index(&self, region_id: u64) -> Option<u64> {
+        if let Some(memtable) = self.get_memtable(region_id) {
+            return memtable.rl().first_index();
         }
+        None
     }
 
-    #[allow(dead_code)]
-    fn purge_expired_files(&self) -> Result<()> {
-        let mut min_file_num = u64::MAX;
-        for memtables in &self.memtables {
-            let memtables = memtables.read().unwrap();
-            let file_num = memtables.values().fold(u64::MAX, |min, x| {
-                cmp::min(min, x.min_file_num().map_or(u64::MAX, |num| num))
-            });
-            if file_num < min_file_num {
-                min_file_num = file_num;
-            }
+    fn last_index(&self, region_id: u64) -> Option<u64> {
+        if let Some(memtable) = self.get_memtable(region_id) {
+            return memtable.rl().last_index();
         }
-
-        self.pipe_log.purge_to(min_file_num)
+        None
     }
 
     fn compact_to(&self, region_id: u64, index: u64) -> u64 {
-        let mut memtables = self.memtables[region_id as usize % SLOTS_COUNT]
-            .write()
-            .unwrap();
-        if let Some(memtable) = memtables.get_mut(&region_id) {
-            return memtable.compact_to(index);
-        }
-        0
+        let first_index = match self.first_index(region_id) {
+            Some(index) => index,
+            None => return 0,
+        };
+
+        let mut log_batch = LogBatch::new();
+        log_batch.add_command(region_id, Command::Compact { index });
+        self.write(log_batch, false).map(|_| ()).unwrap();
+
+        self.first_index(region_id).unwrap_or(index) - first_index
     }
 
     fn compact_cache_to(&self, region_id: u64, index: u64) {
-        let mut memtables = self.memtables[region_id as usize % SLOTS_COUNT]
-            .write()
-            .unwrap();
-        if let Some(memtable) = memtables.get_mut(&region_id) {
-            memtable.compact_cache_to(index);
+        if let Some(memtable) = self.get_memtable(region_id) {
+            memtable.wl().compact_cache_to(index);
         }
     }
 
+    // Write a batch needs 3 steps:
+    // 1. find all involved raft groups and then lock their memtables;
+    // 2. append the log batch to pipe log;
+    // 3. update all involved memtables.
+    // The lock logic is a little complex. However it's necessary because
+    // 1. "Inactive log rewrite" needs to keep logs on pipe log order;
+    // 2. Users can call `append` on one raft group concurrently.
+    // Maybe we can improve the implement of "inactive log rewrite" and
+    // forbid concurrent `append` to remove locks here.
     fn write(&self, log_batch: LogBatch, sync: bool) -> Result<usize> {
+        let mut memtables = Vec::with_capacity(log_batch.items.len());
+        for raft in log_batch.items.iter().map(|item| item.raft_group_id) {
+            memtables.push((raft, self.get_or_insert_memtable(raft)));
```

Comment:
Please get raft ids first, and deduplicate, and then get memtables. Because get memtable will acquire the mutex.